subscription.go

```diff
@@ -31,3 +51,69 @@ func (sub *Subscription) Next(ctx context.Context) (*Message, error) {
 func (sub *Subscription) Cancel() {
 	sub.cancelCh <- sub
 }
+
+func (sub *Subscription) close() {
+	close(sub.ch)
+}
+
+func (sub *Subscription) sendNotification(evt PeerEvent) {
+	sub.evtLogMx.Lock()
+	defer sub.evtLogMx.Unlock()
+
+	sub.addToEventLog(evt)
+}
+
+// addToEventLog assumes a lock has been taken to protect the event log
+func (sub *Subscription) addToEventLog(evt PeerEvent) {
+	e, ok := sub.evtLog[evt.Peer]
+	if !ok {
+		sub.evtLog[evt.Peer] = evt.Type
+		// send signal that an event has been added to the event log
+		select {
+		case sub.evtLogCh <- struct{}{}:
+		default:
+		}
+	} else if e != evt.Type {
+		delete(sub.evtLog, evt.Peer)
+	}
+}
+
+// pullFromEventLog assumes a lock has been taken to protect the event log
+func (sub *Subscription) pullFromEventLog() (PeerEvent, bool) {
+	for k, v := range sub.evtLog {
+		evt := PeerEvent{Peer: k, Type: v}
+		delete(sub.evtLog, k)
+		return evt, true
+	}
+	return PeerEvent{}, false
+}
+
+// NextPeerEvent returns the next event regarding subscribed peers
+// Guarantees: Peer Join and Peer Leave events for a given peer will fire in order.
+// Unless a peer both Joins and Leaves before NextPeerEvent emits either event
+// all events will eventually be received from NextPeerEvent.
+func (sub *Subscription) NextPeerEvent(ctx context.Context) (PeerEvent, error) {
+	for {
+		sub.evtLogMx.Lock()
+		evt, ok := sub.pullFromEventLog()
+		if ok {
+			// make sure an event log signal is available if there are events in the event log
+			if len(sub.evtLog) > 0 {
+				select {
+				case sub.evtLogCh <- struct{}{}:
+				default:
+				}
+			}
```

Comment:
@Stebalien we thinking this works? I should only need to push a signal if there are events in the log right?

Reply:
LGTM.